src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs

```diff
@@ -315,5 +320,36 @@ protected override bool ShouldAnalyzeDeclarationExpression(DeclarationExpression
             // The base analyzer may impose further limitations
             return base.ShouldAnalyzeDeclarationExpression(declaration, semanticModel, cancellationToken);
         }
+
+        private bool IsSwitchExpressionAndCanUseVar(TypeSyntax typeName, ExpressionSyntax initializer, SemanticModel semanticModel)
+        {
+            if (initializer.IsKind(SyntaxKind.SwitchExpression))
+            {
+                // We compare the variable declaration type to each arm's type to see if there is an exact match, or if the
+                // arm type inherits from the variable declaration type. If not, we must use the explicit type instead of var.
+                var declarationType = semanticModel.GetTypeInfo(typeName).Type;
+                if (declarationType != null)
+                {
+                    foreach (var arm in ((SwitchExpressionSyntax)initializer).Arms)
+                    {
+                        var expression = arm.Expression;
+                        if (expression.IsKind(SyntaxKind.ParenthesizedExpression))
+                        {
+                            expression = ((ParenthesizedExpressionSyntax)expression).WalkDownParentheses();
+                        }
+
+                        if (!expression.IsKind(SyntaxKind.ThrowExpression) && !expression.IsKind(SyntaxKind.NullLiteralExpression))
+                        {
+                            var expressionType = semanticModel.GetTypeInfo(expression).Type;
+                            if (expressionType != null && !expressionType.InheritsFromOrEquals(declarationType))
+                            {
+                                return false;
+                            }
```

Comment:
have you consulted a compiler dev about what the rules are about how the "type" of the switch expression is inferred from the arms/destination?  i think we'd need to understand that well to properly code this up.

Reply:
I consulted a compiler dev and was referred to [this document](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/patterns.md#switch-expression). Even after reading through the document, I believe the current fix still stands since it only catches clear cases where the use of 'var' can never be valid. The more intricate cases dealing with implicit conversions are caught by [existing code](https://github.com/dotnet/roslyn/blob/6353783cb067a1d97c1dc577a32f6e1cf497898f/src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs#L292).